Monitoring is a network policy choice (#95)
ietf-wg-scone/scone: draft-ietf-scone-protocol.md

The latter will not be accepted and the former,
even if they do not exceed the path MTU as a result,
can be detected by applications and could be ignored.
This document does not define a mechanism to support detection,
but one might be added in future.

## Monitoring Flows {#monitoring}

Sending throughput advice is optional for any network.
A network that sends throughput advice might, also optionally,
choose to monitor flows
to determine whether applications are following advice.

This section outlines a method
that a network element could use
to determine whether advice is being followed.
Network deployments that choose to monitor
are free to follow any monitoring regime that suits their needs.

This monitoring algorithm is guidance only.
However, monitoring any more strictly than the following
could mean that an application
might be incorrectly classified as not following advice.
A looser monitoring approach,
such as monitoring over a longer time window
than the monitoring period (67s)
or using a higher rate than is signaled,
has no risk of incorrect classification.

When advice is set or updated,
monitoring cannot start until the network element
receives a SCONE packet

> and updates the throughput advice in that packet; see {{apply}}.

Reviewer: ```suggestion
transmits a SCONE packet with
throughput advice other than `Unknown` (127); see {{apply}}.
```

> throughput advice other than `Unknown` (127); see {{apply}}.

Reviewer 2: This paragraph sounds like monitoring devices might throttle flows that don't follow throughput advice from any network device, _including network devices on other hops_. Is that intentional?

I get the intent to describe the algorithm based on what the signaled advice is (because that is what the applications would use in order to comply), however my concern is that if we are to consider such deployments, we might need to expand security considerations how an on-path attacker might abuse SCONE to attack monitoring devices deployed by in other hops.

Reviewer 2: > We can't prevent a network element from monitoring the advice that is contained in packets it's sending,

This is true.

> we can't distinguish between "passing through" advice from upstream and "it set the same rate I was going to set", can we?

From the on-wire signal alone, we can't distinguish those cases, agreed.

However, that is exactly why I think the PR is problematic in its current state.

The phrase "transmits a SCONE packet with 〜" can naturally be read as covering a SCONE signal that originated upstream and is merely being passed through unchanged. And in that case, the following proposed normative text:

> "A network element SHOULD base its monitoring on the maximum value it signaled for throughput advice over at least the preceding monitoring period."

can also be read as telling a monitoring/enforcement device to treat such pass-through / third-party advice as the basis of its monitoring, and by extension, policing / enforcement decisions.

That model does not match the current security model described in the draft, where devices that _act_ on SCONE advice (endpoints) only do so when the SCONE signal is attached to a valid QUIC packet, preventing an off-path attacker from inserting spurious advice. A network element that relies on pass-through / third-party advice cannot apply that "accepts only SCONE signals attached to valid QUIC packets" condition, so text that assumes or endorses that model should not be included in the I-D.

To conclude, I think the draft should talk about the advice that the _hop intends to apply locally,_ rather than whatever SCONE signal that happens to be present in transmitted packets.

Reviewer 3: Relying on packets that "just happen to pass by" is a security breach waiting to happen. The network element has no way to verify that the Scone packet that it sees is genuine. For all we know, an adversary may create a fake Scone packet, coalesced with a fake 1RTT packet. The network element sees that and start its monitoring clock, but the endpoint recognizes the packet as fake and does not heed it. Conflict ensues.

> throughput advice other than `Unknown` (127); see {{apply}}.

Author: We're getting into the meat of this now.

This is the problem: if the network has a policy, then it might suspend enforcement of that policy if it gets a chance to use SCONE.  But we don't need to say that in this particular way.  Let me try again...

> throughput advice other than `Unknown` (127); see {{apply}}.

Reviewer 2: I like this :+1:

Maybe we can tighten the first sentence to: When a network changes the advice, applications need time to adjust their sending behavior.

By omitting "it gives," we avoid the impression that the change starts from when the SCONE packet is sent.

A network element can then monitor affected flows
to determine whether the throughput advice it provided
was followed.

A network element SHOULD base its monitoring
on the maximum value it signaled for throughput advice

> over at least the preceding monitoring period.

Author: ```suggestion
on the maximum value that was configured to apply
during the preceding two monitoring periods.
```
?

A longer period might be used
to account for the possibility that SCONE packets are lost.
This allows applications time to receive advice
and adapt their sending rate.

## Flows That Exceed Throughput Advice {#policing}

A network could deploy policy enforcement that drops or delays packets
to ensure that applications do not exceed throughput limits set in policy.

SCONE allows networks to provide advice to applications,
so that stricter limits,
which can be inefficient and lead to worse application performance,
are not necessary in all cases.

Some applications will not support SCONE.
Other applications either will not
or cannot
follow throughput advice.

Networks can monitor flows to determine if applications follow advice;
see {{monitoring}}.
A network could choose to either disable or loosen policy enforcement
for flows where SCONE is active,
but re-enable or tighten enforcement if monitoring indicates
that throughput advice is not being respected.

Any monitoring and policy enforcement could be implemented

> in different network elements than the ones that signal throughput advice.

Reviewer 2: It might make sense to change like below, to avoid the impression that monitoring and enforcement can be done base on untrusted rate signals (#98):
```suggestion
Monitoring and policy enforcement based on network policy could be implemented
in different network elements than the ones that signal throughput advice.
```

Though, separately, I think this paragraph might need to be moved to #monitoring.

or synthesize datagrams that contain SCONE packets.
The latter will not be accepted and the former,
even if they do not exceed the path MTU as a result,
can be detected by applications and could be ignored.
This document does not define a mechanism to support detection,
but one might be added in future.

## Monitoring Flows {#monitoring}

Sending throughput advice is optional for any network.
A network that sends throughput advice might, also optionally,
choose to monitor flows
to determine whether applications are following advice.

This section outlines a method
that a network element could use
to determine whether advice is being followed.
Network deployments that choose to monitor
are free to follow any monitoring regime that suits their needs.

This monitoring algorithm is guidance only.
However, monitoring any more strictly than the following
could mean that an application
might be incorrectly classified as not following advice.
A looser monitoring approach,
such as monitoring over a longer time window
than the monitoring period (67s)
or using a higher rate than is signaled,
has no risk of incorrect classification.

> Network elements need to wait until advice is signaled.

Reviewer: This reads more like a section title, since the next sentence is describing the waiting too?

When advice is set or updated,

> monitoring cannot start until the network element

Author: Thanks.

receives a SCONE packet

> and updates the throughput advice in that packet; see {{apply}}.

Reviewer: If an element observes advice in a SCONE that is limited - either by the original sender or an upstream element - and does not update it, is it permitted to monitor?

This section outlines a method
that a network element could use
to determine whether advice is being followed.
Network deployments that choose to monitor
are free to follow any monitoring regime that suits their needs.

This monitoring algorithm is guidance only.
However, monitoring any more strictly than the following
could mean that an application
might be incorrectly classified as not following advice.
A looser monitoring approach,
such as monitoring over a longer time window
than the monitoring period (67s)
or using a higher rate than is signaled,
has no risk of incorrect classification.

When a network changes the value
it intends to signal,
applications need time to adjust their sending behavior.
As a result, any monitoring needs to allow time
for SCONE packets to be updated,
for those packets to be received by endpoints,
and for applications to adapt.

A network element can then monitor affected flows
to determine whether the throughput advice it provided
was followed.

A network element SHOULD base its monitoring
on the maximum value that was configured to apply
during the preceding two monitoring periods.
A longer period might be used

> to account for the possibility that SCONE packets are lost.

Reviewer 2: IIRC, we assume 3 SCONE packets to be sent within one monitoring period. Therefore, it seems to me that this advice is not needed for network elements capable of updating every SCONE packet that they forward.

Assuming that is the case, I think we can drop this sentence, or change to the following if we are to close #75 along the way.

```suggestion
If the network element cannot update the throughput advice in every SCONE packet
(or can do so only infrequently), a longer period might be used
to account for the possibility that the updated SCONE packets are lost.
```
